libraries/lib-module-manager/PluginIPCUtils.cpp

```diff
@@ -0,0 +1,157 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file PluginIPCUtils.cpp
+
+  @author Vitaly Sverchinsky
+  
+  Part of lib-module-manager library.
+
+**********************************************************************/
+
+#include "PluginIPCUtils.h"
+
+#include "PluginDescriptor.h"
+
+#include "IPCChannel.h"
+#include "XMLWriter.h"
+
+using namespace details;
+
+namespace {
+   using HeaderBlock = size_t;
+   constexpr auto HeaderBlockSize = sizeof(HeaderBlock);
+
+   constexpr auto NodePlugin = "Plugin";
+   constexpr auto NodeError = "Error";
+   constexpr auto AttrErrorMessage = "msg";
+}
+
+bool details::ParseRequestString(const wxString& req, wxString& providerId, wxString& pluginPath)
+{
+   auto strings = wxSplit(req, ';');
+   if(strings.size() == 2)
+   {
+      providerId = strings[0];
+      pluginPath = strings[1];
+      return true;
+   }
+   return false;
+}
+
+wxString details::MakeRequestString(const wxString& providerId, const wxString& pluginPath)
+{
+   return wxJoin(wxArrayStringEx {providerId, pluginPath}, ';');
+}
+
+void details::PutMessage(IPCChannel& channel, const wxString& value)
+{
+   auto utf8 = value.ToUTF8();
+   const HeaderBlock length = utf8.length();
+   channel.Send(&length, HeaderBlockSize);
+   channel.Send(utf8.data(), length);
+}
+
+void InputMessageReader::ConsumeBytes(const void* bytes, size_t length)
+{
+   auto from = mBuffer.size();
+   mBuffer.resize(from + length);
+   std::memcpy(&mBuffer[from], bytes, length);
+}
+
+bool InputMessageReader::CanPop() const noexcept
+{
+   if(mBuffer.size() >= HeaderBlockSize)
+   {
+      auto expectedSize = *reinterpret_cast<const HeaderBlock*>(mBuffer.data());
+      return mBuffer.size() >= expectedSize + HeaderBlockSize;
+   }
+   return false;
+}
+
+wxString InputMessageReader::Pop()
+{
+   wxString message;
+   const auto length = *reinterpret_cast<const HeaderBlock*>(mBuffer.data());
+   assert(mBuffer.size() >= length + HeaderBlockSize);
+   if(length > 0)
+      message = wxString::FromUTF8(mBuffer.data() + HeaderBlockSize, length);
+   mBuffer.erase(mBuffer.begin(), mBuffer.begin() + HeaderBlockSize + length);
+   return message;
+}
+
+bool PluginValidationResult::IsValid() const noexcept
+{
+   return !mHasError && !mDescriptors.empty();
+}
+
+bool PluginValidationResult::HasError() const noexcept
+{
+   return mHasError;
+}
+
+const wxString& PluginValidationResult::GetErrorMessage() const noexcept
+{
+   return mErrorMessage;
+}
+
+void PluginValidationResult::Add(PluginDescriptor&& desc)
+{
+   mDescriptors.push_back(desc);
```

Comment:
Why not `std::move(desc)`?  You need it because `desc` alone still is an lvalue reference even though the argument is declared with `&&`.

And that might allow vector instead of list for the container, but maybe there is a reason for list I don't understand yet, some requirement of non-relocation of the memory.

Reply:
>Why not std::move(desc)? You need it to avoid copying, because desc alone still is an lvalue reference even though the argument is declared with &&.

Right, thanks!

>And that might allow vector instead of list for the container, but maybe there is a reason for list I don't understand yet, some requirement of non-relocation of the memory.

There is no such requirement, `PluginDescriptor` will be copied later, when added to `PluginManager`